include/npcomp/Dialect/Torch/Transforms/Passes.td

```diff
@@ -0,0 +1,65 @@
+//===-- Passes.td - Pass definition file -------------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef NPCOMP_TORCH_PASSES
+#define NPCOMP_TORCH_PASSES
+
+include "mlir/Pass/PassBase.td"
+
+def GlobalizeObjectGraph : Pass<"torch-globalize-object-graph", "ModuleOp"> {
+  let summary = "Converts TorchScript object graphs to a globalized form";
+  let constructor = "mlir::NPCOMP::Torch::createGlobalizeObjectGraphPass()";
+  let description = [{
+    This pass converts a subset of possible TorchScript modules into a
+    more restrictive lower-level form that strips away the need to be
+    concerned with instances of !torch.nn.Module<...> type. Specifically,
+    the object graph is flattened into a set of discrete globals
+    (`torch.global_slot`) that hold the program state.
+
+    The overarching goal is for a strict correspondence between the original
+    `torch.nn.Module` (call it `root`) that the user `torch.jit.script`'ed, and
+    the public interface of the resulting MLIR module. Specifically:
+      - The call `root.encoder.forward(...)` in Python corresponds to invoking
+        the `func @encoder.forward` on the resulting MLIR module.
+      - The data member access `root.decoder.ids_to_strings_table` in Python
+        corresponds to accessing the
+        `torch.global_slot @decoder.ids_to_strings_table` on the resulting
+        MLIR module.
+    In effect, the entire MLIR module corresponds to an instance of the `root`
+    object. This matches with the intuitive behavior desired for deployment:
```

Comment:
I think this point is arguable, although taking this viewpoint is certainly expedient for now. IREE's execution model can (efficiently) support loading an library once and having multiple session-like "data sections" (i.e. for making parallel requests). A general compilation based on a shared library would be awkward to achieve the same end, though (i.e. would need to use platform specific magic to multi-load the library), and a statically linked executable could not do it at all.

It seems like if you wanted to enable indirect access to the module-attributes, you could model this differently, but you are probably missing a bunch of infra to do that right, and we could look into it later?

Reply:
Good point. I'll keep an eye on it. I feel like our global_slot model here is sufficiently close to IREE's that the "text section" + "data section" transformation should be easy to implement in either place.

Yeah, we could introduce indirect global loads. However, in the general case, it gets really hairy for how to model an object graph with indirection into a flat list of globals (it effectively "denormalizes" the object graph). At that point, I'm inclined to 1) build a great first-class object graph concept in the IR 2) do our best to eliminate it as much as possible 3) but still have a generic runtime "object" type for the general case. I'd just rather *not* frontload "implement a quasi-python object layer in IREE/npcomp", even though I suspect we might eventually need it.

Consider this entire "globalized form" to be provisional. I want to kick the tires on it and see where it falls down in practice, if at all.